lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -678,6 +678,553 @@ LogicalResult ConvertAtenOp<AtenPowTensorScalarOp>::matchAndRewrite(
+    };
+
+    // Transpose needs to done if transposeDims are not non-monotonically
+    // increasing. E.g. [0, 1, 2, 3]: No transpose [1, 0, 2, 3]: Transpose dim0
+    // and dim1 The order need not be sequential, since one or more dims may
+    // have been removed due to broadcasting.
+    auto isTransposeRequired = [](SmallVector<int32_t> transposedDims) -> bool {
+      int32_t lastDim = -1;
+      for (auto &dim : transposedDims) {
+        if (lastDim > dim)
+          return true;
+        lastDim = dim;
+      }
+      return false;
+    };
+
+    SmallVector<TensorShape_t> commonElems, lhsSqueezedElems, rhsSqueezedElems;
+
+    if (!performBatchDimBroadcast) {
+      // Simple with no broadcasting artifacts. Just reshape up to 3D
+      matmulLhs = reshapeUpTo3DTensor(rankBroadcastedLhs);
+      matmulRhs = reshapeUpTo3DTensor(rankBroadcastedRhs);
+
+    } else {
+      // In this case, either or both input matrices involve broadcasting on
+      // their batch dimensions. For example:
+      // 4x5x6, 1x6x7 -> 4x5x7
+      // 4x1x5x6, 1x3x6x7 -> 4x3x5x7
+      // Though maxInputRank is necessarily >=3 here, individual matrices may be
+      // lower rank.
+      // E.g. 3x4x5x6, 6 -> 3x4x5
+
+      // These are the accumulated products of the shape of each dim:
+      // 1. common dimensions: upper dimensions (dims other than two rightmost)
+      // whose shapes are the same for both LHS and RHS.
+      // 2. LHS squeezed dimensions: all dimensions of LHS that involve
+      // broadcasting in either direction, plus the LHS[-2] shape
+      // 3. RHS squeezed dimensions: all dimensions of RHS that involve
+      // broadcasting in either direction, plus the RHS[-1] shape
+      int64_t commonValue = 1, lhsSqueezedValue = 1, rhsSqueezedValue = 1;
+
+      // For both LHS and RHS, the dimensions are separated into the common,
+      // squeezed and remaining dim. E.g. given
+      // LHS = 3x4x5x6
+      // RHS = 1x4x6x7
+      // common = {{dim=1, shape=4}}
+      // lhs squeezed = {{dim=0, shape=3},
+      //                 {dim=2, shape=5}}
+      // rhs squeezed = {{dim=0, shape=1},
+      //                 {dim=2, shape=7}}
+      // The matmul dim is LHS[-1] and RHS[-2], i.e. 6.
+      // Once this is obtained, LHS and RHS are expressed as:
+      // LHS = {common, lhs_squeezed, matmul_dim}
+      // RHS = {common, matmul_dim, rhs_squeezed}
+      // The matmul is then performed to obtain output:
+      // matmul_out = {common, lhs_squeezed, rhs_squeezed}
+      // Finally, we reshape to 'unsqueeze' the LHS and RHS parts and transpose
+      // them back to their correct positions.
+
+      SmallVector<int64_t> transposedLhsShape;
+      SmallVector<int32_t> transposedLhsDims;
+
+      // Step: generate the common dim/shape information
+      for (uint32_t dim = 0; dim < maxInputRank - 2; dim++) {
+        bool isDynamicDim =
+            lhsBroadcastedTy.isDynamic(lhsBroadcastedShape[dim]);
+        if (isDynamicDim ||
```

Comment:
Oh sorry didn't see this comment earlier. 

Dynamic shapes and broadcasting are a complicated topic. Examples:
1. how should a compiler handle broadcasting of 1 vs ? at runtime ? Should '?' be permitted to be any number, in which case the other dim must broadcast to it at runtime and the kernel should accommodate runtime broadcasting. Alternatively the compiler can assert '?' must be 1. 
2. How does runtime handle broadcast of '?' vs '?' ? Should they always match at runtime (no broadcast) ? Or can one of them be 1 (broadcast) ? 

In short, this would require the compiler stack to define these semantics, at which time it can enforce anything needed at compile time. It's for the above reasons that I haven't specifically sought to enforce something here. Hope this answers your question.

Reply:
yea, the compiler stack doesn't fully support broadcasting semantics now. In TorchToLinalg, we would avoid broadcasting for dynamic dims and and runtime assertion to enforce that limitation https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L2360-L2370